design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,206 @@
+# KLIP 9 - Table functions
+
+**Author**: Tim Fox (GitHub: purplefox)
+**Release Target**: 5.4 
+**Status**: In Discussion 
+**Discussion**: https://github.com/confluentinc/ksql/issues/527
+
+## Motivation and background
+
+## Background
+
+A _table valued function_ or _table generating function_ is a concept from the database world.
+It's a function that returns a set of zero or more rows. Contrast this to a scalar function which
+returns a single value. We will call them _table functions_ in the remainder of this document.
+
+Table functions are typically used in FROM clauses with traditional databases. Some databases
+such as PostgreSQL and Hive also allow table functions to appear in SELECT clauses.
+
+An example of a table function from Apache Hive is `explode` - this takes a non scalar type such as
```

Comment:
Are we decided on using `explode` as the name? As far as I know, `explode` arose from looking at Hive as an example and the name just kind of stuck naturally from there. Personally I don't have super strong feelings about `explode` versus `unnest`, which is probably the other most viable option. I find `unnest` more descriptive but I'm fine with either option.

We don't need to spend time on naming this since both options are pretty good, I just thought I'd call out the fact that we do have choices here.

Reply:
Naming is hard! ;)
Personally I find explode more descriptive and I think it evokes better what the function does.
unnesst to me is a bit too much of a software engineering (or maths) term. (and it's not clear to me why unnest means going from one value to zero or more rows) If our target demographic of users is not so technical folks, then my preference would be explode.
Again though, I don't have very strong feelings and we can change easily before release anyway :)

Reply:
Another choice of name which I think is intuitive is `flatten`

Reply:
For me `explode` evokes the idea of expanding one (scalar) value into many values, which isn't exactly what this is doing. I think `unnest` comes from the idea of nested rows, as an array can be considered a row within a row.

Reply:
I know I'm a little bit late with commenting here... just adding my 2 cents by comparing how others have approached this:
* `explode` is also used, e.g. in SparkSQL. BTW they also differentiate between `explode` (just creating multiple rows) vs. `posexplode` (...with indexes/position columns) (see https://spark.apache.org/docs/2.4.0/api/sql/index.html#explode
* interestingly folks at MS have chosen a different approach with ASA and offer CROSS and OUTER APPLY functionality in this regard (see https://docs.microsoft.com/en-us/stream-analytics-query/apply-azure-stream-analytics)